#! /bin/sh /usr/share/dpatch/dpatch-run
## opt-39_noepg.dpatch by Torsten/WarEagle at vdrportal.de
## http://vdrportal.de/board/thread.php?postid=658080#post658080
##
## Thomas Günther <tom@toms-cafe.de>:
##   - solved conflicts with other patches
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: The patch allows to disable normal epg update for specified channels.
## DP: This is useful if you get epg data of the channels from external sources.

@DPATCH@
diff -ur vdr-1.4.7-org/config.c vdr-1.4.7+noEPG/config.c
--- vdr-1.4.7-org/config.c	2007-01-26 14:32:19.000000000 +0100
+++ vdr-1.4.7+noEPG/config.c	2007-10-20 22:30:44.000000000 +0200
@@ -277,1 +277,3 @@
-  InitialVolume = -1;
+  InitialVolume = -1;
+  noEPGMode = 0;
+  noEPGList=strdup("");
@@ -432,1 +434,6 @@
-  else if (!strcasecmp(Name, "ShowReplayMode"))      ShowReplayMode     = atoi(Value);
+  else if (!strcasecmp(Name, "ShowReplayMode"))      ShowReplayMode     = atoi(Value);
+  else if (!strcasecmp(Name, "noEPGMode"))           noEPGMode          = atoi(Value);
+  else if (!strcasecmp(Name, "noEPGList")){
+    free(noEPGList);
+    noEPGList=strdup(Value);
+  }
@@ -506,1 +513,3 @@
-  Store("InitialVolume",      InitialVolume);
+  Store("InitialVolume",      InitialVolume);
+  Store("noEPGMode",          noEPGMode);
+  Store("noEPGList",          noEPGList);
diff -ur vdr-1.4.7-org/config.h vdr-1.4.7+noEPG/config.h
--- vdr-1.4.7-org/config.h	2007-05-12 11:07:16.000000000 +0200
+++ vdr-1.4.7+noEPG/config.h	2007-10-13 18:45:46.000000000 +0200
@@ -254,1 +254,3 @@
-  int InitialVolume;
+  int InitialVolume;
+  char *noEPGList;
+  int noEPGMode;
diff -ur vdr-1.4.7-org/eit.c vdr-1.4.7+noEPG/eit.c
--- vdr-1.4.7-org/eit.c	2006-10-09 18:14:36.000000000 +0200
+++ vdr-1.4.7+noEPG/eit.c	2007-10-13 18:45:46.000000000 +0200
@@ -22,8 +22,28 @@
 class cEIT : public SI::EIT {
 public:
   cEIT(cSchedules *Schedules, int Source, u_char Tid, const u_char *Data, bool OnlyRunningStatus = false);
+private:
+  bool allowedEPG(tChannelID kanalID);
   };
 
+bool cEIT::allowedEPG(tChannelID kanalID){
+  bool rc;
+
+  if (Setup.noEPGMode == 1){
+    rc=false;
+    if (strstr(::Setup.noEPGList,kanalID.ToString())!=NULL){
+      rc=true;
+    }
+  }else{
+    rc=true;
+    if (strstr(::Setup.noEPGList,kanalID.ToString())!=NULL){
+      rc=false;
+    }
+  }
+
+  return rc;
+}
+
 cEIT::cEIT(cSchedules *Schedules, int Source, u_char Tid, const u_char *Data, bool OnlyRunningStatus)
 :SI::EIT(Data, false)
 {
@@ -35,6 +55,15 @@
   if (!channel)
      return; // only collect data for known channels
 
+  // noepg >>>
+  //only use epg from channels not blocked by noEPG-patch
+  tChannelID kanalID;
+  kanalID=channel->GetChannelID();
+  if (!allowedEPG(kanalID)){
+    return;
+  }
+  // noepg <<<
+
   cSchedule *pSchedule = (cSchedule *)Schedules->GetSchedule(channel, true);
 
   bool Empty = true;
diff -ur vdr-1.4.7-org/i18n.c vdr-1.4.7+noEPG/i18n.c
--- vdr-1.4.7-org/i18n.c	2007-02-25 13:08:26.000000000 +0100
+++ vdr-1.4.7+noEPG/i18n.c	2007-10-13 18:45:46.000000000 +0200
@@ -3414,6 +3414,27 @@
     "Tid før EPG skanning (t)",
     "Èasový limit pro snímání EPG (h)",
   },
+  { "Setup.EPG$Mode of noEPG-Patch",
+    "Art des noEPG-Patches",
+    "",
+    "",
+    "",
+    "",
+    "",
+    "",
+    "",
+    "",
+    "",
+    "",
+    "",
+    "",
+    "",
+    "",
+    "",
+    "",
+    "",
+    "",
+  },
   { "Setup.EPG$EPG bugfix level",
     "EPG-Fehlerbereinigung",
     "Nivo za popravilo EPG napak",
diff -ur vdr-1.4.7-org/menu.c vdr-1.4.7+noEPG/menu.c
--- vdr-1.4.7-org/menu.c	2006-12-02 12:12:02.000000000 +0100
+++ vdr-1.4.7+noEPG/menu.c	2007-10-13 18:46:08.000000000 +0200
@@ -2230,6 +2230,7 @@
 
 class cMenuSetupEPG : public cMenuSetupBase {
 private:
+  const char *noEPGModes[2];
   int originalNumLanguages;
   int numLanguages;
   void Setup(void);
@@ -2252,6 +2253,9 @@
 {
   int current = Current();
 
+  noEPGModes[0]=tr("Blacklist");
+  noEPGModes[1]=tr("Whitelist");
+
   Clear();
 
   Add(new cMenuEditIntItem( tr("Setup.EPG$EPG scan timeout (h)"),      &data.EPGScanTimeout));
@@ -2264,1 +2269,2 @@
-     Add(new cMenuEditStraItem(tr("Setup.EPG$Preferred language"),     &data.EPGLanguages[i], I18nNumLanguages, I18nLanguages()));
+     Add(new cMenuEditStraItem(tr("Setup.EPG$Preferred language"),     &data.EPGLanguages[i], I18nNumLanguages, I18nLanguages()));
+  Add(new cMenuEditStraItem(tr("Setup.EPG$Mode noEPG-Patch"),          &data.noEPGMode, 2, noEPGModes));
